### PR TITLE
feat: add variables panel

### PR DESCRIPTION
**Pull request by TrySound**
Ref https://github.com/webstudio-is/webstudio/issues/2532

Here added popover panel to maintain variables. For now only create/edit/delete. Will add bindings support in another PR.

Can be tested in storybook "Variables" component.

Screenshot 2023-11-02 at 02 36 29
Screenshot 2023-11-02 at 02 36 37
<br />
Screenshot 2023-11-02 at 02 37 21
Screenshot 2023-11-02 at 02 38 42

## Code Review

- [ ] hi @kof, I need you to do
  - conceptual review (architecture, feature-correctness)
  - detailed review (read every line)
  - test it on preview

## Before requesting a review

- [ ] made a self-review
- [ ] added inline comments where things may be not obvious (the "why", not "what")

## Before merging

- [ ] tested locally and on preview environment (preview dev login: 5de6)
- [ ] updated [test cases](https://github.com/webstudio-is/webstudio/blob/main/apps/builder/docs/test-cases.md) document
- [ ] added tests
- [ ] if any new env variables are added, added them to `.env.example` and the `builder/env-check.js` if mandatory

### CODE REVIEW

**Comment by istarkov** [comment]
image

Clicking tab I'm visiting this buttons, should be accessed using left-right buttons

File: apps/builder/app/builder/features/settings-panel/variables.tsx [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+import {
+  dataSourcesStore,
+  selectedInstanceSelectorStore,
+} from "~/shared/nano-states";
+import { serverSyncStore } from "~/shared/sync";
+import { CodeEditor } from "./code-editor";
+
+type VariableDataSource = Extract<DataSource, { type: "variable" }>;
+
+// convert value expression to js value
```
Comment (istarkov) [inline apps/builder/app/builder/features/settings-panel/variables.tsx:32]: please use JSDoc comments in that case they are visible on hover in VSCode.
```ts
/**
 *  bla bla bla
 */
```

File: apps/builder/app/builder/features/settings-panel/variables.tsx [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+        return id;
+      },
+    });
+  } catch (error) {
+    result.error = (error as Error).message;
+    return result;
+  }
+  if (ids.size === 0) {
+    try {
+      result.value = eval(code);
```
Comment (istarkov) [inline apps/builder/app/builder/features/settings-panel/variables.tsx:50]: will not work inside cloudlfare?

File: apps/builder/app/builder/features/settings-panel/variables.tsx [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+        </Flex>
+        <Flex justify="end" css={{ gap: theme.spacing[5] }}>
+          <Button color="neutral" onClick={onBack}>
+            Cancel
+          </Button>
+          <Button
+            onClick={() => {
+              if (/[a-zA-z_]+[a-zA-z0-9_]*/.test(name) === false) {
+                setNameErrors([
+                  `Variable name can star with letter or _ and contain letters, digits and underscores`,
```
Comment (istarkov) [inline apps/builder/app/builder/features/settings-panel/variables.tsx:165]: Errors in text star/start, 
can - must
Shoudl be:
`Variable names must start with a letter or underscore (_) and can include numbers and underscores.`

File: apps/builder/app/builder/features/settings-panel/variables.tsx [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+          >
+            Save
+          </Button>
+        </Flex>
+      </Flex>
+    </ScrollArea>
+  );
+};
+
+const selectedInstanceVariables = computed(
```
Comment (istarkov) [inline apps/builder/app/builder/features/settings-panel/variables.tsx:185]: We really need some dom like API for this, so hard to read.
I mean somekind of
`root.queryById(id).getVariables()`
`root.subscribeOnAnyChange`

**Review by istarkov — APPROVED** [review]

**Comment by istarkov** [comment]
ESC in 
image

should work as "Cancel"

**Comment by istarkov** [comment]
- [ ] - Hover on Delete variable tooltip cause hover lost on an item, and prevent to easily hover on next item, imo better to make it unhoverable
image

File: apps/builder/app/builder/features/settings-panel/variables.tsx (hunk 2 above)
Comment (TrySound) [reply apps/builder/app/builder/features/settings-panel/variables.tsx:50]: It's client side code. Variables panel evaluate everything from input

Comment (TrySound) [reply apps/builder/app/builder/features/settings-panel/variables.tsx:50]: Practically json with less verbose syntax

File: apps/builder/app/builder/features/settings-panel/variables.tsx [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+        <Button prefix={<PlusIcon />} onClick={onAdd}>
+          Create variable
+        </Button>
+      </Flex>
+    </Flex>
+  );
+};
+
+const deleteVariable = (variable: VariableDataSource) => {
+  // @todo prevent delete when variable is used
```
Comment (kof) [inline apps/builder/app/builder/features/settings-panel/variables.tsx:221]: probably also needs a confirmation dialog when trying to delete a variable that is used?

File: apps/builder/app/builder/features/settings-panel/variables.tsx [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+    });
+  } catch (error) {
+    result.error = (error as Error).message;
+    return result;
+  }
+  if (ids.size === 0) {
+    try {
+      result.value = eval(code);
+    } catch {
+      result.error = "Unknown error";
```
[the remark this answers is not in the document]
Comment (kof) [reply apps/builder/app/builder/features/settings-panel/variables.tsx:52]: Could be for example: "Parse Error: ${error.message}"

File: apps/builder/app/builder/features/settings-panel/variables.tsx (hunk 5 above)
Comment (TrySound) [reply apps/builder/app/builder/features/settings-panel/variables.tsx:221]: I'd better disable button to let user clear everything self.

**Pull request merged**